vpr/src/place/noc_place_utils.cpp

```diff
@@ -0,0 +1,643 @@
+
+        // we check whether the latency constraint was met here
+        if ((std::max(0., latency - max_latency)) < MIN_EXPECTED_NOC_LATENCY_COST) {
+            count_of_achieved_latency_cons++;
+        }
+    }
+
+    return count_of_achieved_latency_cons;
+}
+
+bool check_recomputed_noc_latency_cost(float recomputed_cost) {
+    return (recomputed_cost < MIN_EXPECTED_NOC_LATENCY_COST) ? false : true;
+}
+
+void allocate_and_load_noc_placement_structs(void) {
+    auto& noc_ctx = g_vpr_ctx.noc();
+
+    int number_of_traffic_flows = noc_ctx.noc_traffic_flows_storage.get_number_of_traffic_flows();
+
+    traffic_flow_aggregate_bandwidth_cost.resize(number_of_traffic_flows, -1);
+    proposed_traffic_flow_aggregate_bandwidth_cost.resize(number_of_traffic_flows, -1);
+    traffic_flow_latency_cost.resize(number_of_traffic_flows, -1);
+    proposed_traffic_flow_latency_cost.resize(number_of_traffic_flows, -1);
+
+    affected_traffic_flows.resize(number_of_traffic_flows, NocTrafficFlowId::INVALID());
+
+    return;
+}
+
+void free_noc_placement_structs(void) {
+    vtr::release_memory(traffic_flow_aggregate_bandwidth_cost);
+    vtr::release_memory(proposed_traffic_flow_aggregate_bandwidth_cost);
+    vtr::release_memory(traffic_flow_latency_cost);
+    vtr::release_memory(proposed_traffic_flow_latency_cost);
+    vtr::release_memory(affected_traffic_flows);
+
+    return;
+}
+
+/* Below are functions related to the feature that forces to the placer to swap router blocks for a certain percentage of the total number of swaps */
+bool check_for_router_swap(int user_supplied_noc_router_swap_percentage) {
+    // generate random number between 0-99 and compare it to the user supplied value
+    return (vtr::irand(99) < user_supplied_noc_router_swap_percentage) ? true : false;
+}
+
+e_create_move propose_router_swap(t_pl_blocks_to_be_moved& blocks_affected, float rlim) {
+    // need to access all the router cluster blocks int he design
+    auto& noc_ctx = g_vpr_ctx.noc();
+    // get a reference to the collection of router cluster blocks in the design
+    const std::unordered_set<ClusterBlockId>& router_clusters = noc_ctx.noc_traffic_flows_storage.get_router_clusters_in_netlist();
+
+    // if there are no router cluster blocks to swap then abort
+    if (router_clusters.empty()) {
+        return e_create_move::ABORT;
+    }
+
+    int number_of_router_blocks = router_clusters.size();
+
+    /* We will choose a random number between 0-number_of_router_blocks-1.
+     * Then we will iterate through the router cluster blocks and stop when we 
+     * have iterated through the chosen random number of blocks. The cluster
+     * we have stopped at will be the cluster to swap.*/
+    int random_cluster_block_index = vtr::irand(number_of_router_blocks - 1);
+    auto router_cluster_block_to_swap_ref = router_clusters.begin();
+    std::advance(router_cluster_block_to_swap_ref, random_cluster_block_index);
```

Comment:
Couldn't this get slow if you have many routers?  I is an O(n) operation on a set, and I believe this data structure is a set.

Reply:
Yes. Will leave as is for now since router count will be small and this utility will eventually be replaced by sara's changes.

Reply:
added to an issue to be worked on later.